verbose error handler can itself throw errors for file uploads

Rob and I discussed this, had an exception raised:

```
2017-11-13 12:12:42,779 [ERROR] cherrypy.error.140684693618304: [13/Nov/2017:12:12:42]  
   Traceback (most recent call last):
      File "/usr/local/uber/env/lib/python3.4/site-packages/cherrypy/_cprequest.py", line 670, in respond
        response.body = self.handler()
      File "/usr/local/uber/env/lib/python3.4/site-packages/cherrypy/lib/encoding.py", line 221, in __call__
        self.body = self.oldhandler(*args, **kwargs)
      File "/usr/local/uber/env/lib/python3.4/site-packages/cherrypy/_cpdispatch.py", line 60, in __call__
        return self.callable(*self.args, **self.kwargs)
      File "/usr/local/uber/plugins/uber/uber/decorators.py", line 382, in with_timing
        return func(*args, **kwargs)
      File "/usr/local/uber/plugins/uber/uber/decorators.py", line 397, in with_session
        retval = func(*args, session=session, **kwargs)
      File "/usr/local/uber/plugins/uber/uber/decorators.py", line 518, in with_restrictions
        return func(*args, **kwargs)
      File "/usr/local/uber/plugins/uber/uber/decorators.py", line 458, in with_rendering
        result = func(*args, **kwargs)
      File "/usr/local/uber/plugins/mivs/mivs/site_sections/mivs_applications.py", line 270, in show_info
        with open(screenshot.filepath, 'wb') as f:
    NameError: name 'screenshot' is not defined
```

Ignore the particulars of that above exception.   The part we're interested in is, our exception handling code (which attempts to log extra information about the request/session/etc) itself failed while parsing the request params with this message:

```

   During handling of the above exception, another exception occurred:
    
   Traceback (most recent call last):
      File "/usr/local/uber/env/lib/python3.4/site-packages/cherrypy/_cprequest.py", line 589, in run
        self.respond(pi)
      File "/usr/local/uber/env/lib/python3.4/site-packages/cherrypy/_cprequest.py", line 690, in respond
        self.handle_error()
      File "/usr/local/uber/env/lib/python3.4/site-packages/cherrypy/_cprequest.py", line 765, in handle_error
        self.hooks.run('before_error_response')
      File "/usr/local/uber/env/lib/python3.4/site-packages/cherrypy/_cprequest.py", line 114, in run
        raise exc
      File "/usr/local/uber/env/lib/python3.4/site-packages/cherrypy/_cprequest.py", line 104, in run
        hook()
      File "/usr/local/uber/env/lib/python3.4/site-packages/cherrypy/_cprequest.py", line 63, in __call__
        return self.callback(**self.kwargs)
      File "/usr/local/uber/plugins/uber/uber/server.py", line 50, in log_exception_with_verbose_context
        log_with_verbose_context('\n'.join([msg, 'Exception encountered']), exc_info=True)
      File "/usr/local/uber/plugins/uber/uber/server.py", line 41, in log_with_verbose_context
        full_msg = '\n'.join([msg, get_verbose_request_context()])
      File "/usr/local/uber/plugins/uber/uber/server.py", line 29, in get_verbose_request_context
        p = ["  %s: %s" % (k, v[:max_reporting_length]) for k, v in cherrypy.request.params.items()]
      File "/usr/local/uber/plugins/uber/uber/server.py", line 29, in <listcomp>
        p = ["  %s: %s" % (k, v[:max_reporting_length]) for k, v in cherrypy.request.params.items()]
    TypeError: 'Part' object is not subscriptable
```

When the error-handling code crashes, that's not great :) .  Luckily, it still prints the original exception (without the extra useful context though).

We should prob write a unit test for our exception handling code to mock and repro this error, and then fix the bug after.

--- a/uber/server.py
+++ b/uber/server.py
@@ -26,7 +26,7 @@ def get_verbose_request_context():
 
     max_reporting_length = 1000   # truncate to reasonably large size in case they uploaded attachments
 
-    p = ["  %s: %s" % (k, v[:max_reporting_length]) for k, v in cherrypy.request.params.items()]
+    p = ["  %s: %s" % (k, str(v)[:max_reporting_length]) for k, v in cherrypy.request.params.items()]
     post_txt = 'Request Params:\n' + '\n'.join(p)
 
     session_txt = 'Session Params:\n' + pformat(cherrypy.session.items(), width=40)

--- a/uber/tests/test_exception_handler.py
+++ b/uber/tests/test_exception_handler.py
@@ -1,0 +1,34 @@
+from uber import config
+from uber.tests import *
+from cherrypy._cpreqbody import Part
+
+
+class MockPart(Part):
+    # fake a cherrypy attachment, make it able to create without constructor
+    def __init__(self):
+        pass
+
+
+class TestExceptionHandler:
+    @pytest.fixture(autouse=True)
+    def setup_cherrypy_fake_env(self, monkeypatch):
+        monkeypatch.setattr(AdminAccount, 'admin_name', Mock(return_value='Bruce'))
+        monkeypatch.setattr(cherrypy.request, 'request_line', "/uber/location3/hello")
+        monkeypatch.setattr(cherrypy.request, 'params', {
+            'id': '32',
+            'action': 'reload',
+            'thing': 3,                 # use a non-string
+            'attachment': MockPart(),   # add fake attachment based on cherrypy's Part() class, make sure we handle OK
+        })
+        monkeypatch.setattr(cherrypy, 'session', {'session_id': '762876'})
+        headers = [
+            ('Content-Type', 'text/html'),
+            ('Server', 'Null CherryPy'),
+            ('Date', localized_now()),
+            ('Content-Length', '80'),
+        ]
+        monkeypatch.setattr(cherrypy.request, 'header_list', headers)
+
+    def test_exception_handler(self):
+        for expected in ['Request', 'text/html', 'reload', 'Bruce', 'session_id', 'Content-Length']:
+            assert expected in server.get_verbose_request_context()

